optimize checks

robieta opened the pull request:
This PR contains several thematically similar micro optimizations:
* Replace `unpack` and `throw_error_for_complex_autograd` with macros. We're currently slinging around compile-time-known strings for the error messages, because C++ isn't good about optimizing that away.
* Kill `unpack_opt`. It is never used.
* `checked_cast_variable` is a relic of the Tensor/Variable merge, and is no longer meaningful.
* Add a simple `Tensor&` overload for `isFwGradDefined` (we're currently wrapping it in an optional, and then unwrapping that in the body), and inline this overload in the header to shave a couple more instructions.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59393**
* #59392
* #59391
* #59389

Commit:
optimize checks

[ghstack-poisoned]

swolchok commented on tools/autograd/templates/VariableType.h:
```diff
@@ -42,14 +42,16 @@ using at::Quantizer;
 
+// TODO: Once we have C++17, this can be a template function rather than a macro.
+#define UNPACK_TENSOR( t, pos ) \
+  TORCH_CHECK( t.defined(), "Expected a Tensor of type Variable but found an undefined Tensor for argument #", pos, " '", #t, "'" ); \
```
use compile-time string literal concatenation -- `" '" #t "'"`

swolchok replied:
if pos is a literal we should stringify it at compile-time too

robieta replied:
Done.

swolchok commented on torch/csrc/autograd/VariableTypeUtils.h:
```diff
@@ -76,13 +76,23 @@ inline void throw_error_out_requires_grad(const char* name) {
+}
+
+inline bool is_complex_autograd(const TensorList& tensorlist) {
+  for (auto tensor: tensorlist) {
```
shouldn't this be `const auto&`?

swolchok commented on torch/csrc/autograd/VariableTypeUtils.h:
```diff
@@ -76,13 +76,23 @@ inline void throw_error_out_requires_grad(const char* name) {
 
+// TODO: Once we have C++17, this can be a template function rather than a macro.
+#define THROW_ERROR_FOR_COMPLEX_AUTOGRAD( t, name ) \
+  TORCH_CHECK( !is_complex_autograd(t), name, " does not support automatic differentiation for outputs with complex dtype." );
```
if name is supposed to be a literal then use compile-time string literal concatenation -- remove the comma

robieta replied:
Done.

albanD commented on torch/csrc/autograd/FunctionsManual.h:
```diff
@@ -32,6 +32,9 @@ struct IndexRangeGenerator {
     size_t i = 0;
 };
 
+inline bool isFwGradDefined(const Tensor& t) {
```
Why is this added? The overload below already covers that no?
Also it would be nice to reduce code in header files...

robieta replied:
The issue is that with just the optional overload we wrap `t` (due to the implicit conversion), and then immediately check that it is defined and then unwrap it on the other side. That turns out to be a non-trivial amount of added work. (Same with inlining; it's a small function on the hot path so we want it to be inlined.)

albanD commented on torch/csrc/autograd/VariableTypeManual.cpp:
```diff
@@ -38,38 +38,7 @@ C10_EXPORT std::vector<at::DeprecatedTypeProperties*> allCUDATypes() {
-}
-}
-
-const Tensor & unpack(const Tensor & t, const char * name, int pos) {
```
Are these functions part of the c++ API? Is this BC-breaking?

robieta replied:
Not to my knowledge. (@ezyang?) If they are, we will still want to make them call the macros and directly use the macros in our own code for perf.

albanD commented on torch/csrc/autograd/VariableTypeUtils.h:
```diff
@@ -76,13 +76,23 @@ inline void throw_error_out_requires_grad(const char* name) {
 }
 
+// TODO: Once we have C++17, this can be a template function rather than a macro.
+#define THROW_ERROR_FOR_COMPLEX_AUTOGRAD( t, name ) \
```
I'm not sure to follow what the benefit of having a macro is here? Since you do more function calls anyways?

albanD commented on tools/autograd/templates/VariableType.h:
```diff
@@ -42,14 +42,16 @@ using at::Quantizer;
 using c10::optional;
 
+// TODO: Once we have C++17, this can be a template function rather than a macro.
+#define UNPACK_TENSOR( t, pos ) \
```
Do we really need a macro for this given that it is only used in the codegen anyways? You can replace the macro in the template with these two lines.

This used to be much more complex before the Tensor/Variable merge, but not much is left.

robieta replied:
I elected for the macro because I didn't want the manually defined functions to drift from the code-gen'd ones.

albanD replied:
They should be updated the same way. But it is know that whenever we change the codegen, the same update should be done on the manually defined functions.

albanD commented on tools/autograd/gen_inplace_or_view_type.py:
```diff
@@ -170,13 +173,10 @@ def unpack_args(f: NativeFunction) -> Tuple[List[str], List[Binding]]:
-        ref = (not is_nullable) and not is_tensor_list
-        suffix = '_opt' if is_nullable and not is_tensor_list else ''
-        body.append(UNPACK_TENSOR.substitute(
+        template = (UNPACK_TENSOR_LIST if is_tensor_list else UNPACK_TENSOR)
```
I'm not sure how this is equivalent to the old code, in particular, I don't see the `_list` magic there.
Did you diff the generated code to make sure it is still the same?

robieta replied:
I didn't. (I'll make a note to do so.) list magic is now handled by using a different codegen template. The key point for understanding equivalence is that there are no instances of unpack_opt, so the logic simplifies.

Commit:
Update on "optimize checks"

This PR contains several thematically similar micro optimizations:
* Replace `unpack` and `throw_error_for_complex_autograd` with macros. We're currently slinging around compile-time-known strings for the error messages, because C++ isn't good about optimizing that away.
* Kill `unpack_opt`. It is never used.
* `checked_cast_variable` is a relic of the Tensor/Variable merge, and is no longer meaningful.
* Add a simple `Tensor&` overload for `isFwGradDefined` (we're currently wrapping it in an optional, and then unwrapping that in the body), and inline this overload in the header to shave a couple more instructions.



[ghstack-poisoned]

Commit:
Update on "optimize checks"

This PR contains several thematically similar micro optimizations:
* Replace `unpack` and `throw_error_for_complex_autograd` with macros. We're currently slinging around compile-time-known strings for the error messages, because C++ isn't good about optimizing that away.
* Kill `unpack_opt`. It is never used.
* `checked_cast_variable` is a relic of the Tensor/Variable merge, and is no longer meaningful.
* Add a simple `Tensor&` overload for `isFwGradDefined` (we're currently wrapping it in an optional, and then unwrapping that in the body), and inline this overload in the header to shave a couple more instructions.



[ghstack-poisoned]

Commit:
Update on "optimize checks"

This PR contains several thematically similar micro optimizations:
* Replace `unpack` and `throw_error_for_complex_autograd` with macros. We're currently slinging around compile-time-known strings for the error messages, because C++ isn't good about optimizing that away.
* Kill `unpack_opt`. It is never used.
* `checked_cast_variable` is a relic of the Tensor/Variable merge, and is no longer meaningful.
* Add a simple `Tensor&` overload for `isFwGradDefined` (we're currently wrapping it in an optional, and then unwrapping that in the body), and inline this overload in the header to shave a couple more instructions.



[ghstack-poisoned]

Commit:
Update on "optimize checks"

This PR contains several thematically similar micro optimizations:
* Replace `unpack` and `throw_error_for_complex_autograd` with macros. We're currently slinging around compile-time-known strings for the error messages, because C++ isn't good about optimizing that away.
* Kill `unpack_opt`. It is never used.
* `checked_cast_variable` is a relic of the Tensor/Variable merge, and is no longer meaningful.
* Add a simple `Tensor&` overload for `isFwGradDefined` (we're currently wrapping it in an optional, and then unwrapping that in the body), and inline this overload in the header to shave a couple more instructions.



[ghstack-poisoned]

Commit:
Update on "optimize checks"

This PR contains several thematically similar micro optimizations:
* Replace `unpack` and `throw_error_for_complex_autograd` with macros. We're currently slinging around compile-time-known strings for the error messages, because C++ isn't good about optimizing that away.
* Kill `unpack_opt`. It is never used.
* `checked_cast_variable` is a relic of the Tensor/Variable merge, and is no longer meaningful.
* Add a simple `Tensor&` overload for `isFwGradDefined` (we're currently wrapping it in an optional, and then unwrapping that in the body), and inline this overload in the header to shave a couple more instructions.



[ghstack-poisoned]

Commit:
Update on "optimize checks"

This PR contains several thematically similar micro optimizations:
* Replace `unpack` and `throw_error_for_complex_autograd` with macros. We're currently slinging around compile-time-known strings for the error messages, because C++ isn't good about optimizing that away.
* Kill `unpack_opt`. It is never used.
* `checked_cast_variable` is a relic of the Tensor/Variable merge, and is no longer meaningful.
* Add a simple `Tensor&` overload for `isFwGradDefined` (we're currently wrapping it in an optional, and then unwrapping that in the body), and inline this overload in the header to shave a couple more instructions.



[ghstack-poisoned]

Commit:
Update on "optimize checks"

This PR contains several thematically similar micro optimizations:
* Replace `unpack` and `throw_error_for_complex_autograd` with macros. We're currently slinging around compile-time-known strings for the error messages, because C++ isn't good about optimizing that away.
* Kill `unpack_opt`. It is never used.
* `checked_cast_variable` is a relic of the Tensor/Variable merge, and is no longer meaningful.
* Add a simple `Tensor&` overload for `isFwGradDefined` (we're currently wrapping it in an optional, and then unwrapping that in the body), and inline this overload in the header to shave a couple more instructions.



[ghstack-poisoned]

Commit:
Update on "optimize checks"

This PR contains several thematically similar micro optimizations:
* Replace `unpack` and `throw_error_for_complex_autograd` with macros. We're currently slinging around compile-time-known strings for the error messages, because C++ isn't good about optimizing that away.
* Kill `unpack_opt`. It is never used.
* `checked_cast_variable` is a relic of the Tensor/Variable merge, and is no longer meaningful.
* Add a simple `Tensor&` overload for `isFwGradDefined` (we're currently wrapping it in an optional, and then unwrapping that in the body), and inline this overload in the header to shave a couple more instructions.



[ghstack-poisoned]

Commit:
Update on "optimize checks"

This PR contains several thematically similar micro optimizations:
* Replace `unpack` and `throw_error_for_complex_autograd` with macros. We're currently slinging around compile-time-known strings for the error messages, because C++ isn't good about optimizing that away.
* Kill `unpack_opt`. It is never used.
* `checked_cast_variable` is a relic of the Tensor/Variable merge, and is no longer meaningful.
* Add a simple `Tensor&` overload for `isFwGradDefined` (we're currently wrapping it in an optional, and then unwrapping that in the body), and inline this overload in the header to shave a couple more instructions.



[ghstack-poisoned]

Commit:
Update on "optimize checks"

This PR contains several thematically similar micro optimizations:
* Replace `unpack` and `throw_error_for_complex_autograd` with macros. We're currently slinging around compile-time-known strings for the error messages, because C++ isn't good about optimizing that away.
* Kill `unpack_opt`. It is never used.
* `checked_cast_variable` is a relic of the Tensor/Variable merge, and is no longer meaningful.
* Add a simple `Tensor&` overload for `isFwGradDefined` (we're currently wrapping it in an optional, and then unwrapping that in the body), and inline this overload in the header to shave a couple more instructions.



[ghstack-poisoned]

Commit:
Update on "optimize checks"

This PR contains several thematically similar micro optimizations:
* Replace `unpack` and `throw_error_for_complex_autograd` with macros. We're currently slinging around compile-time-known strings for the error messages, because C++ isn't good about optimizing that away.
* Kill `unpack_opt`. It is never used.
* `checked_cast_variable` is a relic of the Tensor/Variable merge, and is no longer meaningful.
* Add a simple `Tensor&` overload for `isFwGradDefined` (we're currently wrapping it in an optional, and then unwrapping that in the body), and inline this overload in the header to shave a couple more instructions.



[ghstack-poisoned]

Commit:
Update on "optimize checks"

This PR contains several thematically similar micro optimizations:
* Replace `unpack` and `throw_error_for_complex_autograd` with macros. We're currently slinging around compile-time-known strings for the error messages, because C++ isn't good about optimizing that away.
* Kill `unpack_opt`. It is never used.
* `checked_cast_variable` is a relic of the Tensor/Variable merge, and is no longer meaningful.
* Add a simple `Tensor&` overload for `isFwGradDefined` (we're currently wrapping it in an optional, and then unwrapping that in the body), and inline this overload in the header to shave a couple more instructions.



[ghstack-poisoned]

Commit:
Update on "optimize checks"

This PR contains several thematically similar micro optimizations:
* Replace `unpack` and `throw_error_for_complex_autograd` with macros. We're currently slinging around compile-time-known strings for the error messages, because C++ isn't good about optimizing that away.
* Kill `unpack_opt`. It is never used.
* `checked_cast_variable` is a relic of the Tensor/Variable merge, and is no longer meaningful.
* Add a simple `Tensor&` overload for `isFwGradDefined` (we're currently wrapping it in an optional, and then unwrapping that in the body), and inline this overload in the header to shave a couple more instructions.



[ghstack-poisoned]

Commit:
Update on "optimize checks"

This PR contains several thematically similar micro optimizations:
* Replace `unpack` and `throw_error_for_complex_autograd` with macros. We're currently slinging around compile-time-known strings for the error messages, because C++ isn't good about optimizing that away.
* Kill `unpack_opt`. It is never used.
* `checked_cast_variable` is a relic of the Tensor/Variable merge, and is no longer meaningful.
* Add a simple `Tensor&` overload for `isFwGradDefined` (we're currently wrapping it in an optional, and then unwrapping that in the body), and inline this overload in the header to shave a couple more instructions.



[ghstack-poisoned]

Commit:
Update on "optimize checks"

This PR contains several thematically similar micro optimizations:
* Replace `unpack` and `throw_error_for_complex_autograd` with macros. We're currently slinging around compile-time-known strings for the error messages, because C++ isn't good about optimizing that away.
* Kill `unpack_opt`. It is never used.
* `checked_cast_variable` is a relic of the Tensor/Variable merge, and is no longer meaningful.
* Add a simple `Tensor&` overload for `isFwGradDefined` (we're currently wrapping it in an optional, and then unwrapping that in the body), and inline this overload in the header to shave a couple more instructions.



[ghstack-poisoned]

Commit:
Update on "optimize checks"

This PR contains several thematically similar micro optimizations:
* Replace `unpack` and `throw_error_for_complex_autograd` with macros. We're currently slinging around compile-time-known strings for the error messages, because C++ isn't good about optimizing that away.
* Kill `unpack_opt`. It is never used.
* `checked_cast_variable` is a relic of the Tensor/Variable merge, and is no longer meaningful.
* Add a simple `Tensor&` overload for `isFwGradDefined` (we're currently wrapping it in an optional, and then unwrapping that in the body), and inline this overload in the header to shave a couple more instructions.



[ghstack-poisoned]

Commit:
Update on "optimize checks"

This PR contains several thematically similar micro optimizations:
* Replace `unpack` and `throw_error_for_complex_autograd` with macros. We're currently slinging around compile-time-known strings for the error messages, because C++ isn't good about optimizing that away.
* Kill `unpack_opt`. It is never used.
* `checked_cast_variable` is a relic of the Tensor/Variable merge, and is no longer meaningful.
* Add a simple `Tensor&` overload for `isFwGradDefined` (we're currently wrapping it in an optional, and then unwrapping that in the body), and inline this overload in the header to shave a couple more instructions.



[ghstack-poisoned]

Commit:
Update on "optimize checks"

This PR contains several thematically similar micro optimizations:
* Replace `unpack` and `throw_error_for_complex_autograd` with macros. We're currently slinging around compile-time-known strings for the error messages, because C++ isn't good about optimizing that away.
* Kill `unpack_opt`. It is never used.
* `checked_cast_variable` is a relic of the Tensor/Variable merge, and is no longer meaningful.
* Add a simple `Tensor&` overload for `isFwGradDefined` (we're currently wrapping it in an optional, and then unwrapping that in the body), and inline this overload in the header to shave a couple more instructions.



[ghstack-poisoned]

Commit:
Update on "optimize checks"

This PR contains several thematically similar micro optimizations:
* Replace `unpack` and `throw_error_for_complex_autograd` with macros. We're currently slinging around compile-time-known strings for the error messages, because C++ isn't good about optimizing that away.
* Kill `unpack_opt`. It is never used.
* `checked_cast_variable` is a relic of the Tensor/Variable merge, and is no longer meaningful.
* Add a simple `Tensor&` overload for `isFwGradDefined` (we're currently wrapping it in an optional, and then unwrapping that in the body), and inline this overload in the header to shave a couple more instructions.



[ghstack-poisoned]

Commit:
Update on "optimize checks"

This PR contains several thematically similar micro optimizations:
* Replace `unpack` and `throw_error_for_complex_autograd` with macros. We're currently slinging around compile-time-known strings for the error messages, because C++ isn't good about optimizing that away.
* Kill `unpack_opt`. It is never used.
* `checked_cast_variable` is a relic of the Tensor/Variable merge, and is no longer meaningful.
* Add a simple `Tensor&` overload for `isFwGradDefined` (we're currently wrapping it in an optional, and then unwrapping that in the body), and inline this overload in the header to shave a couple more instructions.



[ghstack-poisoned]

Commit:
Update on "optimize checks"

This PR contains several thematically similar micro optimizations:
* Replace `unpack` and `throw_error_for_complex_autograd` with macros. We're currently slinging around compile-time-known strings for the error messages, because C++ isn't good about optimizing that away.
* Kill `unpack_opt`. It is never used.
* `checked_cast_variable` is a relic of the Tensor/Variable merge, and is no longer meaningful.
* Add a simple `Tensor&` overload for `isFwGradDefined` (we're currently wrapping it in an optional, and then unwrapping that in the body), and inline this overload in the header to shave a couple more instructions.



[ghstack-poisoned]

Commit:
Update on "optimize checks"

This PR contains several thematically similar micro optimizations:
* Replace `unpack` and `throw_error_for_complex_autograd` with macros. We're currently slinging around compile-time-known strings for the error messages, because C++ isn't good about optimizing that away.
* Kill `unpack_opt`. It is never used.
* `checked_cast_variable` is a relic of the Tensor/Variable merge, and is no longer meaningful.
* Add a simple `Tensor&` overload for `isFwGradDefined` (we're currently wrapping it in an optional, and then unwrapping that in the body), and inline this overload in the header to shave a couple more instructions.



[ghstack-poisoned]

Commit:
Update on "optimize checks"

This PR contains several thematically similar micro optimizations:
* Replace `unpack` and `throw_error_for_complex_autograd` with macros. We're currently slinging around compile-time-known strings for the error messages, because C++ isn't good about optimizing that away.
* Kill `unpack_opt`. It is never used.
* `checked_cast_variable` is a relic of the Tensor/Variable merge, and is no longer meaningful.
* Add a simple `Tensor&` overload for `isFwGradDefined` (we're currently wrapping it in an optional, and then unwrapping that in the body), and inline this overload in the header to shave a couple more instructions.



[ghstack-poisoned]
